README.md
Extension templates generator
=============================

Plugin for Cotonti CMF. 
Allow to fast generate template (files) for Cotonti Extension for further development.

Description
-----------

It's simply creates standard set of Extension files with recommended folder structure and 
necessary file format. It's maked to help Extension developers fast create plugins and 
not to waste the time for filling standard parameters.

Features
--------

* Both of Cotonti branch are supported (Templates for Genoa and Siena versions)
* Highly configurable through templates and parameters
* Unlimited parts for plugin supported
* Creates folder structure and files with Cotonti development guidelines
* Addition files like `readme.md` description created

Requirements
------------

Works with Cotonti Siena (0.9.x branch). No addition requirements.

Demo page
---------

No demo page available. Install it to use.


### Comments

Original idea and most part of code was born several years ago for Genoa. So some part of code are ugly.
Some refactoring are done just to convert it for Siena.


### How extension works

Main controls is accessible after install in Admin section of plugin 

To create a simple pack of plugin files just select Cotonti branch, enter plugin name and
check these files you are needed. Then press «Create Extension template files» button.
Addition setting you can perform using «Section 2: Misc» button.
Generated files (you need write permission for that) can be found in folder configured in 
Setting menu of plugin, or «created» subfolder in plugin folder.


Install
-------

* Unpack, copy files to root folder of your site.
* Install via Admin → Extensions menu (`Administration panel → Extensions`)
* Configure some attributes on Plugin Config section.

### Comments

After install you can change settings (author name and e-mail) then goto Admin → Tools page
to work with it (`Administration panel → Extensions → Extension template → Administration`).

References
----------

* [Cotonti.com](http://Cotonti.com/) -- Home of Cotonti CMF
* [ExtTemplate](https://github.com/macik/cot-extension_template) -- «Extension Template» on GitHub



CHANGELOG.md
Extension template generator 
============================

`Extension template generator` is an plugin for Cotonti CMF 
to generate boilerplate files of Cotonti Extension for further development.

Changelog
---------

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/) and 
CHANGELOG [formatting rules](http://keepachangelog.com/).


## v1.3.1 (2016-Jul-10)
* change versions log file to more CHANGELOG compliant style
* fixed setup conf to meet 0.9.19 version 

## v1.3.0 (2016-Mar-10)  
* Siena plugin templates reworked (simplified view with use of snippets)

## v1.2.5 (2014-Aug-31)


## v1.2.0 (2013-Aug-28)  

* Fixed extension generation in Genoa mode
+ Added proper workflow with different admin themes (Priori, Bootstrap, Foster)
+ Generation log view without settings reset
* Fixed working with ajax on/off
+ Javascript separated from HTML block 
* Javascript rewritten for common jQuery version compatibility 
* Some minor fixes in Genoa templates

## v1.1.2 (2013-Mar-15)  

- Removed closing PHP tags from Siena templates
+ Support for new style of naming Extension categories array (since Siena 0.9.13)

## v1.1.1 (2013-Feb-15)  

+ Option to select plugin category
+ Option to select placement of MD files
* Some small fixes


## v1.1.0 (2013-Jan-31)

* Refactoring
- Removed jQueryUI dependency
+ Addintion templates
+ Adding MD files in created templates


## v1.0.3 (2012-Jan-27)

* Some improvements in templates
+ Adding TPL and CSS files in created templates


## v1.0.1 (2012-Oct-16)

+ Fixes for Siena prior to 0.9.10


## v1.0.0 (2012-Sep-21)

* Refactoring to use in Siena


## v0.7 (2011-Aug-17)

+ Added ability to create Siena plugins
+ Added templates for Siena branch


## v0.6 (2010-Oct-11)

* Some styles for admin part
* Moved to admin part


## v0.5 (2009-Oct-28)

+ Added simle templates
+ Added Log for created files


## v0.4 (2009-Mar-15)

* Base version for Genoa



plugins/extension_template/extension_template.ajax.php
<?php
/* ====================
[BEGIN_COT_EXT]
Hooks=ajax
[END_COT_EXT]
==================== */

/**
 * Ajax file for Extension template plugin
 *
 * @package extension_template
 * @author Andrey Matsovkin
 * @copyright Copyright (c) 2008-2013
 * @license Distributed under BSD License.
 */

if (!defined('COT_CODE') && !defined('COT_PLUG')) { die('Wrong URL'); }

define('extension_template_AJAX',TRUE);
$plug_name = 'extension_template';
$ajax_link = cot_url('plug',array('r'=>$plug_name));
list($usr['auth_read'], $usr['auth_write'], $usr['isadmin']) = cot_auth('plug', $r);
cot_block($usr['isadmin']);

header("Content-type: text/html; charset=utf-8");
if ( $_SERVER['HTTP_X_REQUESTED_WITH'] == 'XMLHttpRequest' || (strtolower($_SERVER['SERVER_NAME'])=='localhost') ) {
	if ($mode) {
		require_once cot_incfile($plug_name, 'plug');
		$count =0;
		foreach ($tpl_arr as $k => $v) { // отрисовка форм параметров
			$count++;
			$in['num']=$count;
			$in['val']=str_replace('"',"'",$v);
			$in['text']= (in_array($k,$bold_tag)) ? "<b>$k</b>" : $k;
			$tpl -> assign('in',$in);
			$tpl->parse('PRM.INPUT_PRM');
		}
		$tpl->parse('PRM');
		echo $tpl->text('PRM');
	} else {
		echo '';
	}
} else {
    echo 'Test result. NaAQ (not an ajax query)';
}


plugins/extension_template/extension_template.setup.php
<?php
/* ====================
[BEGIN_COT_EXT]
Code=extension_template
Name=Extension template
Category=development-maintenance
Description=Generates sample (template) files for extension development
Version=1.3.0
Date=2016-mar-10
Author=Andrey Matsovkin
Copyright=Copyright (c) 2008-2013, Andrey Matsovkin
Notes=If your enjoy my plugin please consider donating to help support future developments. mailto: macik.spb@gmail.com   Thanks!
Auth_guests=
Lock_guests=W2345A
Auth_members=
Lock_members=12345
[END_COT_EXT]

[BEGIN_COT_EXT_CONFIG]
version=01:string::1.0.0:Initial version of plugin
author=02:string::Author Name:Author name
email=03:string::author_email@example.com:Author E-mail
licence=04:select:MIT,GPL,BSD,CC-BY-SA:BSD:Licence type
outputdir=05:string::datas\created\:Folder for output files
md_outside=06:radio::0:Place *.md files outside plugin folder
[END_COT_EXT_CONFIG]
==================== */

/**
 * Extension template plugin for Cotonti CMF
 *
 * @package extension_template
 * @author Andrey Matsovkin
 * @copyright Copyright (c) 2008-2013
 * @license Distributed under BSD License.
*/

defined('COT_CODE') or header('HTTP/1.1 '.$m='403 Forbidden') . die($m);




plugins/extension_template/lang/extension_template.en.lang.php
<?php
/**
 * Localization file for Extension template
 * @author Andrey Matsovkin
 * @copyright Copyright (c) 2008-2013
 * @license Distributed under BSD License.
 */

defined('COT_CODE') or die('Wrong URL');

$L['plu_title'] = 'Extension templates'; //
$plug_title = 'Extension template generator ';

$L['mplug_part'] = 'Section';
$L['mplug_stage1'] = 'Section 1: Main';
$L['mplug_stage2'] = 'Section 2: Misc';
$L['mplug_name'] = 'Plugin name';
$L['mplug_category'] = 'Plugin category';
$L['mplug_name2'] = '(with one word, required only A..Z letters and unerscore symbol )';
$L['mplug_lang'] = 'Create LANG files (2 file for ru,en)';

$L['mplug_main'] = 'Extension «standalone» file (plugin_name.php)';
$L['mplug_setup'] = 'Setup file (plugin_name.setup.php)';
$L['mplug_addon'] = 'Addition files';
$L['mplug_goto2'] = 'Create Extension template files';

$L['mplug_size'] = 'Total files size {$0} bytes';
$L['mplug_targetdir'] = 'Plugin files located in `{$0}` folder.';
$L['mplug_log'] = 'Wrote {$0} bytes in file <b>{$1}</b>';
$L['mplug_creating'] = 'Files creation log';
$L['mplug_err1'] = 'Error creating folder: {$1}';
$L['mplug_back'] = 'Back to configuration and creates another one template';
$L['mplug_params'] = 'Check Extension params';

$L['mplug_select'] = 'Select target version of CMF';
$L['mplug_folders'] = 'Make addition folder';
$L['mplug_subfolders'] = '(comma separated subfolders)';
$L['mplug_langdesc'] = '(comma separated lang codes, e.g.: ru,en)';
$L['mplug_common'] = 'Shared extension files ';

$L['mplug_mdfiles'] = 'Other files ';
$L['mplug_mddesc'] = 'Description file for plugin';
$L['mplug_mdvers'] = 'Сhanges log file';

$L['mplug_error'] = 'Error: ';
$L['mplug_nofolder'] = 'Error: Folder "{$folder}" not exists';
$L['mplug_nofile'] = 'Error: File "{$file}" not exists';
$L['mplug_newfolder'] = 'Creating new folder: {$1}';
$L['mplug_cantcreate'] = 'Error: Can not create Folder "{$folder}". Check write permission for it.';

$L['mplug_nojquery'] = 'Sorry, this extension works only with javascript enables and jQuery library. You can turn it on on <a href="'.cot_url('admin','m=config&n=edit&o=core&p=performance').'" title="Performance settings">settings page</a>.';

$L['mplug_wrongmode'] = 'Wrong Mode settings: ';

$L['info_desc'] ='Generates Extension files (template) for further development';

$L['cfg_author'] =array('Author of plugin');
$L['cfg_email'] =array('E-mail of author');
$L['cfg_licence'] =array('Licence type','you can change it in parameters while creating template');
$L['cfg_md_outside'] =array('Place *.md files outside plugin folder','if selected it will be places in `created` folder');




plugins/extension_template/lang/extension_template.ru.lang.php
<?php
/**
 * Localization file for Extension template
 * @author Andrey Matsovkin
 * @copyright Copyright (c) 2008-2013
 * @license Distributed under BSD License.
 */

defined('COT_CODE') or die('Wrong URL');

$plug_title = 'Создание шаблона плагина ';
$L['plu_title'] = 'Создание шаблона плагина';

$L['mplug_part'] = 'Раздел';
$L['mplug_stage1'] = 'Раздел 1: Выбор названия плагина и файлов для создания';
$L['mplug_stage2'] = 'Раздел 2: Выбор параметров плагина';
$L['mplug_name'] = 'Название плагина';
$L['mplug_name2'] = '(в одно слово, латинскими буквами, допустим символ "_")';
$L['mplug_category'] = 'Категория плагина';
$L['mplug_lang'] = 'Создать файлы языковых настроек';

$L['mplug_main'] = 'Основной файл плагина (plugin_name.php)';
$L['mplug_setup'] = 'Файл установки и настройки плагина (plugin_name.setup.php)';
$L['mplug_addon'] = 'Дополнительные файлы плагина';
$L['mplug_goto2'] = 'Создать шаблон-макет плагина';

$L['mplug_size'] = 'Общий размер файлов {$0} байт';
$L['mplug_targetdir'] = 'Созданные файлы находятся в каталоге `{$0}`.';
$L['mplug_log'] = 'Создан файл <b>{$1}</b> ({$0} байт)';
$L['mplug_creating'] = 'Отчет о создании файлов';
$L['mplug_err1'] = 'Ошибка при создании каталога {$1}';
$L['mplug_back'] = 'Вернуться к выбору параметров и создать еще один плагин';
$L['mplug_params'] = 'Уточните параметры для шаблона';

$L['mplug_select'] = 'Выберите версию Cotonti для которой необходимо создать файлы';
$L['mplug_folders'] = 'Создать дополнительные каталоги';
$L['mplug_subfolders'] = '(перечислите через запятую названия вложенных каталогов)';
$L['mplug_langdesc'] = '(перечислите через запятую обозначения языков, например: ru,en)';
$L['mplug_common'] = 'Общие файлы плагина';

$L['mplug_mdfiles'] = 'Дополнительные файлы ';
$L['mplug_mddesc'] = 'Файл Markdown с описанием плагина';
$L['mplug_mdvers'] = 'Лог изменений';

$L['mplug_error'] = 'Ошибка: ';
$L['mplug_nofolder'] = 'Ошибка: Каталог "{$folder}" не найден';
$L['mplug_nofile'] = 'Ошибка: Файл "{$file}" не найден';
$L['mplug_newfolder'] = 'Создан каталог: {$1}';
$L['mplug_cantcreate'] = 'Ошибка: Не могу создать каталог "{$folder}". Проверьте права на запись.';

$L['mplug_wrongmode'] = 'Ошибка выбора режима: ';

$L['mplug_nojquery'] = 'Извините, данный плагин работает только при включенном Javascript и библиотеке jQuery. Включить её вы можете в меню <a href="'.cot_url('admin','m=config&n=edit&o=core&p=performance').'" title="Настройки производительности">настроек</a>.';

$L['info_desc'] ='Генератор шаблонных файлов (заготовки) для разработки Расширения'; // plugin description

$L['cfg_author'] =array('Автор плагина');
$L['cfg_email'] =array('Эл.адрес автора');
$L['cfg_licence'] =array('Тип лицензии','может быть изменен в параметрах при создании шаблона');
$L['cfg_md_outside'] =array('Создавать *.md файлы вне каталога плагина','если опция выбрана файлы будут созданы рядом с папкой плагина, в каталоге `created`');




plugins/extension_template/inc/extension_template.functions.php
<?php
/**
 * Extension template plugin API
 * @package extension_template
 * @author Andrey Matsovkin
 * @copyright Copyright (c) 2008-2013
 * @license Distributed under BSD License.
 */

defined('COT_CODE') or die('Wrong URL');
require_once cot_langfile('extension_template', 'plug');
$plug_name = 'extension_template';

/**
 * Returns Extension Categories. Support new naming convention
 * used since Cotonti v0.9.13
 */
function getExtensionCategories(){
	global $L;
	if (isset($L['ext_cat']) && is_array($L['ext_cat'])) {
		return $L['ext_cat'];
	} else {
		$ext_cat = array();
		foreach ($L as $k => $v) {
			if (substr($k,0,8)=='ext_cat_') $ext_cat[substr($k,8)] = $v;
		}
		return $ext_cat;
	}

}
/**
 * Parses text for key => value pairs and makes Array
 * @param str $text - text to pare
 * @param str $delimiter - key-value delimiter
 * @param str $linebreak - lines delimiter
 */
function parseParams($text,$delimiter='>>',$linebreak=PHP_EOL){
	$arr = explode($linebreak, $text);
	$tpl_arr = array();
	foreach ($arr as $value) {
		list($mrk,$str) = explode($delimiter, $value);
		$tpl_arr[trim($mrk)]=trim($str);
	}
	return $tpl_arr;
}

/**
 * Check for proper template file depends on Name and Mode and return
 * initialized XTemplate
 * @param string $mode - Mode for templates
 * @param string $name - Name of template
 * @param string $type - Default prefix for template name
 * @return XTemplate|boolean
 */
function getTemplate($mode,$name,$type='make_'){
	global $cfg, $plug_name;
	if (empty($name)) return false;
	$tpl_filename = $cfg['plugins_dir']."/$plug_name/tpl/$mode/{$type}{$name}.tpl";
	if (file_exists($tpl_filename)) return new XTemplate($tpl_filename);
	return false;
}

/**
 * Copies file from $source to $target with log prints
 * @param string $source - source file
 * @param string $target - target file
 * @param string $text_to_insert - text to insert before file contents
 */
function copyFile($source,$target,$text_to_insert=''){
	if (file_exists($source)) {
		$file = pathinfo($target,PATHINFO_FILENAME).'.'.pathinfo($target,PATHINFO_EXTENSION);
		if ($text_to_insert) {
			toLog(cot_rc('mplug_log',array(sizeCounter(file_put_contents($target, $text_to_insert.file_get_contents($source))), $file)));
		} elseif (copy($source,$target)) {
			toLog(cot_rc('mplug_log',array(sizeCounter(filesize($target)), $file)));
		}
	} else {
		toLog(cot_rc('mplug_nofile',array('file'=>$source)));
	}
}

/**
 * Creates file with template
 * Uses global $ex_tpl template if other not specified
 * @param string $filename - Filename (with path) to create file
 * @param XTemplate $tpl - Template
 * @return integer|boolean - Number of bytes written or false otherwise
 */
function makeFile($filename,$tpl=null){
	global $ex_tpl, $tpl_arr, $L;
	if ($filename && !is_null($tpl)) {
		$file = pathinfo($filename,PATHINFO_FILENAME).'.'.pathinfo($filename,PATHINFO_EXTENSION);
		$tpl->assign($tpl_arr);
		$tpl->parse('MAIN');
		toLog(cot_rc('mplug_log',array(sizeCounter(file_put_contents($filename,$tpl->text('MAIN'))),$file)));
		return true;
	}
	return false;
}

/**
 * Adds line to Log output
 * @param string $str
 */
function toLog($str=''){
	global $log_tpl;
	if ($str) {
		$log_tpl->assign('logline',$str);
		$log_tpl->parse('LOG.LINE');
	}
	return $str;
}

/**
 * Add Error msg line to Log output
 * @param string $msg - Folder name cause Error
 */
function toError($msg=''){
	global $log_tpl;
	if ($str) {
		$log_tpl->assign('err_msg',cot_rc('mplug_err1','err='.$msg));
		$log_tpl->parse('LOG.ERROR');
	}
}

/**
 * Check Folder is exists and trying to create it otherwise.
 * @param string $dir_path Path to folder
 * @return boolean - TRUE on folder exists or created, FALSE otherwise.
 */
function checkDir($dir_path){
	global $cfg;
	$dirs = explode('/',$dir_path);
	$dirname = array_pop($dirs);
	if (!$dirname) $dirname = array_pop($dirs);
	return file_exists($dir_path)
	|| ( @mkdir($dir_path, $cfg['dir_perms'], true) && toLog(cot_rc('mplug_newfolder',array(1=>$dirname))));
}

/**
 * counts total files size
 * @param number $size - size of file
 * @return size of file
 */
function sizeCounter($size=0){
	global $extpl_total_size, $extpl_last_size;
	if ($size) {
		$extpl_last_size = $size;
		$extpl_total_size += $size;
	}
	return $size;
}

$mode = cot_import('mode','G','ALP') ? cot_import('mode','G','ALP') : (cot_import('plugtyp','P','ALP') ? cot_import('plugtyp','P','ALP') : 'SIENA');

$pfname = 'plug_filename';

$tpl = new XTemplate(cot_tplfile($plug_name, 'plug'));

// setting up different params depend on mode (Genoa version or Siena)
$tpl->parse('TPL_MODES');
$tpl_modes = parseParams($tpl->text('TPL_MODES'));
if (!array_key_exists($mode, $tpl_modes)) {
	echo $L['mplug_wrongmode'].$mode;
	exit;
}
$mode_block = 'TPL_MARKS_'.strtoupper($mode);
// init array of tokens for selected mode (version)
$tpl->parse($mode_block);
$tpl_arr = parseParams($tpl->text($mode_block ));

$bold_tag = array('MMP_PLUGNAME','MMP_PLUGDATE','MMP_VARPREFIX','MMP_PLUGFILE','MMP_HOOK','MMP_PLUGPART','MMP_TAGS','MMP_PLUGTITLE','MMP_PLUGDESC');
$dimtag = array('');

$mode= strtolower($mode);



plugins/extension_template/extension_template.tools.php
<?php
/* ====================
[BEGIN_COT_EXT]
Hooks=tools
[END_COT_EXT]
==================== */

/**
 * Administration panel for Extension template
 *
 * @package extension_template
 * @author Andrey Matsovkin
 * @copyright Copyright (c) 2008-2013
 * @license Distributed under BSD License.
 */

defined('COT_CODE') or header('HTTP/1.1 '.$m='403 Forbidden') . die($m);

/*
 * TODO: Extended plugin
*  ? save settings as profiles
*  + add Forbidden ext names list
*  + add controls to set setup file variables
*  + add support to make modules
*  + add links to add addition parts (ajax, tools, rc)
*  + close accordion on «generate» press
*
**/

$plug_name = 'extension_template';
$base_path = $cfg['plugins_dir']."/$plug_name";
$ext_name_re = '/[^a-z0-9_\-]+/i';

switch ($cfg['admintheme']) {
	case 'foster':
	case 'bootstrap':
		$et_thememode = 1;
		break;
	case 'priori':
		$et_thememode = 2;
		break;
}

list($usr['auth_read'], $usr['auth_write'], $usr['isadmin']) = cot_auth('plug', $plug_name);
cot_block($usr['isadmin']);

$scriptname = cot_url('admin','m=other&p=extension_template');

$base_path = $cfg['plugins_dir']."/$plug_name";
$snippets = $cfg['plugins_dir'].'/extension_template/tpl/siena/snippets';
$inc_path  = "$base_path/inc";
$tpl_path  = "$base_path/tpl";

$TAG = '<?php';
$ajax_link = (function_exists('cot_url')) ? cot_url('index',"r=$plug_name") : "plug.php?r=$plug_name";
$et_cfg = $cfg['plugin'][$plug_name];
//XTemplate::init(false); // debug only
require_once cot_incfile($plug_name, 'plug');

$forbidden_module_names = array(
'mysqlport', 'mysqldb', 'mysqlcharset', 'mysqlcollate', 'mainurl',
'multihost', 'site_id', 'secret_key', 'defaulttheme', 'defaultscheme',
'defaulticons', 'defaultlang', 'enablecustomhf', 'cache', 'cache_drv',
'xtpl_cache', 'cache_index', 'cache_page', 'cache_forums', 'check_updates',
'display_errors', 'redirmode', 'xmlclient', 'ipcheck', 'authcache',
'customfuncs', 'new_install', 'avatars_dir', 'cache_dir', 'lang_dir',
'modules_dir', 'pfs_dir', 'photos_dir', 'plugins_dir', 'system_dir',
'thumbs_dir', 'themes_dir', 'extrafield_files_dir', 'icons_dir', 'dir_perms',
'file_perms', 'debug_mode', 'debug_logpath', 'shield_force',
'useremailduplicate', 'version', 'dbversion', 'forcedefaultlang',
'defaulttimezone', 'adminemail', 'clustermode', 'hostip', 'maintenance',
'maintenancereason', 'devmode', 'parser', 'redirbkonlogin', 'redirbkonlogout',
'maxrowsperpage', 'easypagenav', 'confirmlinks', 'jquery', 'turnajax',
'topline', 'banner', 'bottomline', 'menu1', 'menu2', 'menu3', 'menu4',
'menu5', 'menu6', 'menu7', 'menu8', 'menu9', 'freetext1', 'freetext2',
'freetext3', 'freetext4', 'freetext5', 'freetext6', 'freetext7', 'freetext8',
'freetext9', 'gzip', 'headrc_consolidate', 'headrc_minify', 'jquery_cdn',
'cookiedomain', 'cookiepath', 'cookielifetime', 'captchamain',
'captcharandom', 'forcedefaulttheme', 'homebreadcrumb', 'separator',
'disablesysinfos', 'keepcrbottom', 'showsqlstats', 'msg_separate',
'maintitle', 'subtitle', 'metakeywords', 'title_users_details',
'title_header', 'title_header_index', 'subject_mail', 'body_mail',
'forcerememberme', 'timedout', 'plugin', 'referercheck', 'shieldzhammer',
'shieldtadjust', 'shieldenabled', 'hashfunc', 'default_show_installed',
	);
$forbidden_plugin_names = array(
	''
);


switch ($a) {
	case 'step2': // generates files with given params
		$log_tpl = new XTemplate(cot_tplfile($plug_name, 'plug'));

		$keys = array_keys($tpl_arr);
		foreach ($_POST as $k => $v) { // gets all params starts with PRM
			if (substr($k,0,3)=='prm') {
				// need some escape magic for NOTE parameter (ID:#12)
				$val = cot_import($k,'P',($k == 'prm_12') ? 'HTM' : 'TXT');
				if ($k == 'prm_12') $val = str_replace("'","\'",$val);
				$tpl_arr[$keys[substr($k,4)-1]] = $val;
			}
		}
		$plf = cot_import('plf','P','ARR');//$_POST['plf'];
		$tpl_arr['MMP_PLUGNAME'] = preg_replace($ext_name_re, '', $tpl_arr['MMP_PLUGNAME']);
		$base_folder = $et_cfg['outputdir'] ? $et_cfg['outputdir'] : $cfg['plugins_dir']."/$plug_name/created/";
		$base_folder = str_replace('\\','/',$base_folder);
		$base_folder .= $base_folder[strlen($base_folder)-1]=='/' ? '' : '/';
		$main_folder = $base_folder.$tpl_arr['MMP_PLUGNAME'].'/'; // for ex. plugins/extension_template/created/plugname/

		if (checkDir($main_folder)) { // makes Extension base folder
			// copying plugin sample icon
			$source = $cfg['plugins_dir']."/$plug_name/img/icon.png";
			$target = $main_folder.$tpl_arr['MMP_PLUGNAME'].'.png';
			if ($mode=='siena') copyFile($source, $target);

			if ( $plf['makedirs']) { // if subfolders needed // $mode=='siena' &&
				$dirs = explode(',',$plf['dirsname']);
/*				if ($mode=='cot') {  проверить назначение
					array_push($dirs,'setup');
				}  */
				$dirs = array_unique($dirs);
				foreach ($dirs as $dirname) {
					if (checkDir($main_folder.$dirname)) {
						$source_folder = $cfg['plugins_dir']."/$plug_name/tpl/$mode/$dirname/";
						//toLog(cot_rc($L['mplug_newfolder'],array(1=>$dirname)));
						if ($dirname == 'tpl') {
							if ($mode=='siena') {
								// copying css file
								$source = $source_folder.'make.css';
								$target = "$main_folder$dirname/{$tpl_arr['MMP_PLUGNAME']}.css";
								copyFile($source, $target);
								// copying tpl sample file
								$source = $cfg['plugins_dir']."/$plug_name/tpl/$mode/tpl/make.tpl";
								$target = "$main_folder$dirname/{$tpl_arr['MMP_PLUGNAME']}.tpl";
								copyFile($source, $target, "# main TPL file for {$tpl_arr['MMP_PLUGTITLE']} plugin\n");
							}
						}
					} else {
						toLog(cot_rc('mplug_cantcreate',array(1=>$main_folder.$dirname)));
					}
				}
			}
			if ($plf['lang']) { // for lang files
				$lng_folder = $main_folder.'lang/';
				if (checkDir($lng_folder)) {
					$langs = explode(',',$plf['langlist']);
					foreach ($langs as $lng) { // for languages
						if (preg_match("/[A-Z]{2}/i", $lng)) {
							$ex_tpl = getTemplate($mode,'lang');
							$tpl_arr['MMP_PLUGLANG'] = $lng;
							$lng_filename =$tpl_arr['MMP_PLUGNAME'].'.'.$lng.'.lang.php';
							makeFile($lng_folder.$lng_filename, $ex_tpl);
						}
					}
				} else toError($lng_folder);
			}
			$plug_files = array('main','setup'); // makes main files
			foreach ($plug_files as $part_name) {
				if ($plf[$part_name]) {
					$ex_tpl = getTemplate($mode, $part_name);
					$ex_tpl->assign($tpl_arr);
					if ($part_name == 'setup') {
						$ex_tpl->parse('MAIN.CONF_VAR');
					}
					$file_name =$tpl_arr['MMP_PLUGNAME'].(($part_name=='main')?'':'.'.$part_name).'.php';
					makeFile($main_folder.$file_name, $ex_tpl);
				}
			}
			$plug = cot_import('plug','P','ARR');
			foreach ($plug as $k1 => $v1) { // makes addintion files with related templates
				if ($v1['used']) {
					$pl_file[$k1]['name'] = $v1['name'];
					$tplname = strtolower($v1['name']);
					if (preg_match("/ajax/i",$tplname)) {
						$tplname = 'ajax';
					}
					$tpl_arr['MMP_PLUGFILE'] = $tpl_arr['MMP_PLUGNAME'].'.'.$v1['name'];
					$tpl_arr['MMP_PLUGPART'] = $v1['name'];

					// if we have own template for certain filename
					$ex_tpl = getTemplate($mode, $tplname);
					if (!$ex_tpl) { // otherway use dump tpl with custom hook
						$tpl_arr['MMP_TAGS'] = '';
						$tpl_arr['MMP_HOOK'] = strtolower($tplname).PHP_EOL;
						$ex_tpl = getTemplate($mode, 'dump');
					}

					$file_name = $tpl_arr['MMP_PLUGNAME'].'.'.$v1['name'].'.php';
					makeFile($main_folder.$file_name, $ex_tpl);
				}
			}

			if ($mode=='siena' && is_array($plf['siena'])) { // makes Siena specified files
				foreach ($plf['siena'] as $k=>$v) {
					if ($v) {
						$ex_tpl = getTemplate($mode, $k, '');
						if ($ex_tpl) {
							$res_file= $tpl_arr['MMP_PLUGNAME'].'.'.$k;
							if (in_array($k[0],array('c','i','u')) ) {// these files we placed in «setup» folde
								$folder = $main_folder.'setup';
							} else {
								$folder = $main_folder.'inc';
							}
							if (checkDir($folder)) {
								makeFile($folder.'/'.$res_file, $ex_tpl);
							}
						}
					}
				}
			}
			$md_folder = $et_cfg['md_outside'] ? $base_folder : $main_folder;
			if (is_array($plf['misc'])) {
				foreach ($plf['misc'] as $k=>$v) {
					if ($v) {
						if ($k == 'readme.md') {
							if (!is_array($langs)) $langs[] = 'en';
							foreach ($langs as $lng) { // for languages
								$lng = strtolower($lng);
								if (preg_match("/[A-Z]{2}/i", $lng)) {
									$k = 'README'.($lng!='en'?'_'.$lng:'').'.md';
									$ex_tpl = getTemplate('misc', $k, '');
									makeFile($md_folder.$k, $ex_tpl);
								}
							}
							continue;
						}
						$ex_tpl = getTemplate('misc', $k, '');
						makeFile($md_folder.$k, $ex_tpl);
					}
				}
			}
		} else {
			toError($main_folder);
		}
		toLog(cot_rc('mplug_size',array($extpl_total_size)));
		toLog(cot_rc('mplug_targetdir',array($main_folder.$dirname)));
		$log_tpl->parse('LOG');
		if ($cfg['turnajax']) {
			echo $log_tpl->text('LOG');
			exit;
		} else {
			$plugin_body = $log_tpl->text('LOG');
		}
		break;

	default: // generate form with extension params
		if ($cfg['jquery']) {
			$count =0;
			$ext_cat = getExtensionCategories();
			$ext_cat_selector = cot_selectbox('', 'plf[ext_cat]', array_keys($ext_cat),array_map(html_entity_decode,array_values($ext_cat)),true);
			foreach ($tpl_modes as $k => $v) { // Mode (version) selection: Genoa / Siena
				$count++;
				$in['id']=$count;
				$in['typ']=$k;
				$in['desc']=$v;
				$in['check']= ($count==2) ? ' checked="checked"': '';
				$tpl -> assign('in',$in);
				$tpl->parse('MAIN.JQUERY.INPUT_TYP');
			}
			$count =0;
			foreach ($tpl_arr as $k => $v) { // отрисовка форм параметров
				$count++;
				$in['num']=$count;
				$in['val']=str_replace('"',"'",$v);
				$in['text']= (in_array($k,$bold_tag)) ? "<b>$k</b>" : $k;
				$tpl -> assign('in',$in);
				$tpl->parse('PRM.INPUT_PRM');
			}
			$tpl->parse('PRM');
			$tpl->assign('PRM',$tpl->text('PRM'));
			$tpl->parse('MAIN_CFG');
			$tpl->assign('MAIN_CFG',$tpl->text('MAIN_CFG'));
			$tpl->parse('MAIN.JQUERY');
		} else {
			$tpl->parse('MAIN.NOJQUERY');
		}
		$tpl->parse('MAIN');
		$plugin_body .= $tpl->text('MAIN');
		break;
}

